### ISSUE TITLE
Inconsistent resolution of dependencies and generation of lockfile

### ISSUE THREAD

**Issue by 1natsu172** [pnpm/pnpm#8678]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### Which area(s) of pnpm are affected? (leave empty if unsure)

Lockfile

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/1natsu-vacation/weird-dependencies-dedupe-pnpm-repro/tree/main

### Reproduction steps

1. `corepack enable`
2. `pnpm rm @webext-core/messaging`
  - Clean lockfile(reproduce initialization)
4. `pnpm add @webext-core/messaging`
  - Added only info about the `@webext-core/messaging` in lockfile
  - ```
    snapshots:

      '@webext-core/messaging@2.0.1': {}
    ```
5. `pnpm add @webext-core/messaging` (Add again)
  - Added more info about dependencies in lockfile
  - part of e.g.
    - ```
      .
      .
      .
      snapshots:

        '@lukeed/csprng@1.1.0': {}

        '@webext-core/messaging@2.0.1':
          dependencies:
            serialize-error: 11.0.3
            uid: 2.0.2
            webextension-polyfill: 0.10.0
      .
      .
      .
      ```
6. `pnpm dedupe --check`
  - Duplicates found, suggesting first-time installation status
  - ```
    ERR_PNPM_DEDUPE_CHECK_ISSUES  Dedupe --check found changes to the lockfile

    Packages
    @webext-core/messaging@2.0.1
    ├── - serialize-error 11.0.3
    ├── - uid 2.0.2
    └── - webextension-polyfill 0.10.0
    
    - @lukeed/csprng@1.1.0
    - serialize-error@11.0.3
    - type-fest@2.19.0
    - uid@2.0.2
    - webextension-polyfill@0.10.0
    
    Run pnpm dedupe to apply the changes above.
    ```
7. `pnpm dedupe`
  - Return to status 4(first-time installation lockfile)
8. `pnpm i --fix-lockfile`
  - Return to status 5 (second-time installation lockfile)


### Describe the Bug

## List of Issues

- Dependency modules are not installed correctly the first time installation
  - After two installations(again install with same command), these are installed correctly.
- Dedupe is performed, but it's not correct
- `--fix-lockfile` return it to the correct status.

## Description

The dependency modules of the library([`@webext-core/messaging`](https://github.com/aklinker1/webext-core/tree/main/packages/messaging)) is not written correctly in the lockfile. The library has three dependencies(`serialize-error`, `uid`, `webextension-polyfill`), but the first install does not write them to the lockfile.(even `node_modules` does not come with those dependencies!).

**But when I execute the install again, the dependency modules are installed as expected. The lock file is also updated.** Then, when dedupe is performed after the expected results are achieved, the dependency modules are lost. And furthermore, execute `pnpm i --fix-lockfile` returns to the expected status. 

In other words, the sequence of generation is not consistent.

_※There may be other libraries where this occurs, but I'm not sure._



### Expected Behavior

- Dependency modules installed and lock files generated correctly on the first install.
- No duplicates are detected when checking for dedupe.
- No differences are detected after `--fix-lockfile`

### Which Node.js version are you using?

22.10.0

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

**Comment by 1natsu172** [comment]
Close because it may have been a registry problem, not pnpm.
